File: test_solver.py
import unittest
from solver import *


class Testgrid(unittest.TestCase):

    def test_initialize_grid(self):
        rows = 6
        cols = 6
        grid = initialize_grid(rows, cols)
        self.assertEqual(len(grid), rows)
        self.assertEqual(len(grid[0]), cols)
        self.assertEqual(len(grid[0][0]), 4)
        self.assertEqual(grid[0][0][0], -1)

    def test_initialize_pieces(self):
        pieces, _ = initialize_pieces(puzzles_format=VANSTONE_FORMAT, filename='bvanston_puzzles/Data6_2_6/RandPuzzle06x06_100.txt')
        self.assertEqual(tuple(pieces[0]), (0, 1, 2, 0))
        self.assertEqual(tuple(pieces[1]), (0, 1, 1, 0))
        self.assertEqual(tuple(pieces[2]), (0, 2, 1, 0))
        self.assertEqual(tuple(pieces[3]), (0, 2, 2, 0))

    def test_pieces_to_orientations(self):
        pieces, _ = initialize_pieces(puzzles_format=VANSTONE_FORMAT, filename='bvanston_puzzles/Data6_2_6/RandPuzzle06x06_100.txt')
        pieces =  pieces_to_orientations(pieces)
        self.assertEqual(len(pieces), 6 * 6 * 4)

        self.assertEqual(tuple(pieces[0]), (0, 1, 2, 0))
        self.assertEqual(tuple(pieces[1]), (0, 0, 1, 2))

    def test_rotate_piece(self):
        piece = (1, 2, 3, 4)
        self.assertEqual(rotate_piece(piece, 0), (1, 2, 3, 4))
        self.assertEqual(rotate_piece(piece, 1), (4, 1, 2, 3))
        self.assertEqual(rotate_piece(piece, 2), (3, 4, 1, 2))
        self.assertEqual(rotate_piece(piece, 3), (2, 3, 4, 1))

class TestAction(unittest.TestCase):

    def test_place_piece_on_grid(self):
        grid = initialize_grid(6, 6)
        rows = 6
        cols = 6
        piece = (GRAY, 2, 3, 4)

        success, grid, next_position = place_piece_on_grid(grid, piece, (0, 1))

        self.assertEqual(success, True)
        self.assertEqual(grid[0][1][0], GRAY)
        self.assertEqual(grid[0][1][1], 2)
        self.assertEqual(grid[0][1][2], 3)
        self.assertEqual(grid[0][1][3], 4)
        self.assertEqual(next_position, (0, 2))

    def test_get_next_position(self):
        grid = initialize_grid(6, 6)
        piece = (1, 2, 3, 4)
        next_position = get_next_position(grid, (1, 1))
        self.assertEqual(
            get_next_position(grid, (1, 1)),
            (1, 2)
        )

    def test_get_next_position_circular(self):
        grid = initialize_grid(3, 3)
        piece = (1, 2, 3, 4)
        next_position = get_next_position(grid, (1, 1), is_circular=True)
        self.assertEqual(
            get_next_position(grid, (0, 0), is_circular=True),
            (0, 1)
        )

        self.assertEqual(
            get_next_position(grid, (0, 1), is_circular=True),
            (0, 2)
        )

        self.assertEqual(
            get_next_position(grid, (0, 2), is_circular=True),
            (1, 2)
        )

        self.assertEqual(
            get_next_position(grid, (1, 2), is_circular=True),
            (2, 2)
        )

        self.assertEqual(
            get_next_position(grid, (2, 2), is_circular=True),
            (2, 1)
        )

        self.assertEqual(
            get_next_position(grid, (2, 1), is_circular=True),
            (2, 0)
        )

        self.assertEqual(
            get_next_position(grid, (2, 0), is_circular=True),
            (1, 0)
        )

        self.assertEqual(
            get_next_position(grid, (1, 0), is_circular=True),
            (1, 1)
        )
    def test_get_next_position_circular_4_4(self):
        grid = initialize_grid(4, 4)
        piece = (1, 2, 3, 4)
        next_position = get_next_position(grid, (1, 1), is_circular=True)
        self.assertEqual(
            get_next_position(grid, (1, 0), is_circular=True),
            (1, 1)
        )
        self.assertEqual(
            get_next_position(grid, (1, 1), is_circular=True),
            (1, 2)
        )
        self.assertEqual(
            get_next_position(grid, (1, 2), is_circular=True),
            (2, 1)
        )
        self.assertEqual(
            get_next_position(grid, (2, 1), is_circular=True),
            (2, 2)
        )

    def test_is_move_legal_borders(self):
        grid = initialize_grid(6, 6)
        piece = (GRAY, 3, 2, GRAY)
        rows = 6
        self.assertEqual(is_move_legal(grid, piece, (0, 0)), True)

        piece = (2, 3, 2, 2)
        self.assertEqual(is_move_legal(grid, piece, (0, 0)), False)

        piece = (GRAY, 2, 2, 3)
        self.assertEqual(is_move_legal(grid, piece, (0, 0)), False)

        piece = (3, 3, 2, GRAY)
        self.assertEqual(is_move_legal(grid, piece, (0, 0)), False)

        piece = (3, 3, 2, GRAY)
        self.assertEqual(is_move_legal(grid, piece, (1, 0)), True)
        
        piece = (3, 3, GRAY, 3)
        self.assertEqual(is_move_legal(grid, piece, (1, 0)), False)

        piece = (GRAY, 2, GRAY, GRAY)
        self.assertEqual(is_move_legal(grid, piece, (1, 0)), False)

        piece = (3, 3, GRAY, 3)
        self.assertEqual(is_move_legal(grid, piece, (5, 0)), False)

        piece = (3, GRAY, GRAY, 3)
        self.assertEqual(is_move_legal(grid, piece, (5, 5)), True)

        piece = (3, GRAY, 2, 3)
        self.assertEqual(is_move_legal(grid, piece, (5, 5)), False)

        piece = (3, 2, GRAY, 3)
        self.assertEqual(is_move_legal(grid, piece, (5, 5)), False)

    def test_border_tile_cannot_be_in_center(self):
        grid = initialize_grid(6, 6)
        piece = (GRAY, 3, 2, 3)
        self.assertEqual(is_move_legal(grid, piece, (1, 1)), False)
        
        piece = (2, 3, GRAY, 3)
        self.assertEqual(is_move_legal(grid, piece, (1, 1)), False)

        piece = (2, 3, 4, GRAY)
        self.assertEqual(is_move_legal(grid, piece, (1, 1)), False)

        piece = (2, GRAY, 4, 2)
        self.assertEqual(is_move_legal(grid, piece, (1, 1)), False)

    def test_border_tiles_neighbours(self):
        grid = initialize_grid(6, 6)
        grid[0][0][NORTH] = GRAY
        grid[0][0][SOUTH] = 2
        grid[0][0][WEST] = GRAY
        grid[0][0][EAST] = 1

        piece = (GRAY, 3,  2, 1)
        self.assertEqual(is_move_legal(grid, piece, (0, 1)), True)

        piece = (GRAY, 3, 2, 2)
        self.assertEqual(is_move_legal(grid, piece, (0, 1)), False)

        piece = (2, 3, 2, GRAY)
        self.assertEqual(is_move_legal(grid, piece, (1, 0)), True)

        piece = (3, 3, 2, GRAY)
        self.assertEqual(is_move_legal(grid, piece, (1, 0)), False)

File: test_solver.cpp
#define CATCH_CONFIG_MAIN  // This tells Catch to provide a main() - only do this in one cpp file
#include "catch.hpp"
#include "solver.hpp"
using namespace std;

TEST_CASE("Rotate piece", "[factorial]") {
    std::array<int, 4> piece = {1, 2, 3, 4}; 
    REQUIRE(rotate_piece(piece, 1)[0] == 4);
    REQUIRE(rotate_piece(piece, 1)[1] == 1);
    REQUIRE(rotate_piece(piece, 1)[2] == 2);
    REQUIRE(rotate_piece(piece, 1)[3] == 3);

    REQUIRE(rotate_piece(piece, 2)[0] == 3);
    REQUIRE(rotate_piece(piece, 2)[1] == 4);
    REQUIRE(rotate_piece(piece, 2)[2] == 1);
    REQUIRE(rotate_piece(piece, 2)[3] == 2);

    REQUIRE(rotate_piece(piece, 3)[0] == 2);
    REQUIRE(rotate_piece(piece, 3)[1] == 3);
    REQUIRE(rotate_piece(piece, 3)[2] == 4);
    REQUIRE(rotate_piece(piece, 3)[3] == 1);
}

TEST_CASE("Get next position", "[factorial]") {
    int cols = 6; 
    std::array<int, 4> piece = {1, 2, 3, 4}; 
    std::array<int, 2> position = {1, 1};

    std::array<int, 2> next_position = get_next_position(cols, position);
    REQUIRE(next_position[0] == 1);
    REQUIRE(next_position[1] == 2);

    position = {1, 2};
    next_position = get_next_position(cols, position);
    REQUIRE(next_position[0] == 1);
    REQUIRE(next_position[1] == 3);

    position = {1, 3};
    next_position = get_next_position(cols, position);
    REQUIRE(next_position[0] == 1);
    REQUIRE(next_position[1] == 4);

    position = {1, 4};
    next_position = get_next_position(cols, position);
    REQUIRE(next_position[0] == 1);
    REQUIRE(next_position[1] == 5);

    position = {1, 5};
    next_position = get_next_position(cols, position);
    REQUIRE(next_position[0] == 2);
    REQUIRE(next_position[1] == 0);

    position = {3, 5};
    next_position = get_next_position(cols, position);
    REQUIRE(next_position[0] == 4);
    REQUIRE(next_position[1] == 0);
}

TEST_CASE("Initialize grid", "[factorial]") {
    int cols = 6; 
    int rows = 6; 

    std::vector<int> grid = initialize_grid(rows, cols);
    REQUIRE(grid[0 * cols * 4 + 0 * 4 + 0] == EMPTY);
    REQUIRE(grid[5 * cols * 4 + 5 * 4 + 3] == EMPTY);
    REQUIRE(grid[5 * cols * 4 + 5 * 4 + 3] == EMPTY);
}

TEST_CASE("Is move legal", "[factorial]") {
    int cols = 6; 
    int rows = 6; 

    std::vector<int> grid = initialize_grid(rows, cols);
    std::array<int, 4> piece = {1, 2, 3, 4}; 
    std::array<int, 2> position = {0, 0};
    REQUIRE(is_move_legal(grid, piece, position, 6, 6) == false);
    std::array<int, 4> piece_2 = {0, 1, 2, 0}; 
    REQUIRE(is_move_legal(grid, piece_2, position, 6, 6) == true);
    std::array<int, 4> piece_3 = {0, 2, 0, 0}; 
    REQUIRE(is_move_legal(grid, piece_3, position, 6, 6) == false);
}

TEST_CASE("Is move legal 2", "[factorial]") {
    int cols = 6; 
    int rows = 6; 

    std::vector<int> grid = initialize_grid(rows, cols);
    std::array<int, 4> piece = {0, 4, 1, 0}; 
    std::array<int, 2> position = {0, 0};

    std::vector<int> new_grid;
    std::array<int, 2> new_position;
    bool success;

    std::tie(success, new_grid, new_position) = place_piece_on_grid(grid, piece, position, rows, cols);
    REQUIRE(success == true);
    REQUIRE(new_grid[0] == 0);
    REQUIRE(new_grid[1] == 4);
    REQUIRE(new_grid[2] == 1);
    REQUIRE(new_grid[3] == 0);
    REQUIRE(new_position[0] == 0);
    REQUIRE(new_position[1] == 1);

    piece = {0, 2, 1, 3}; 
    REQUIRE(is_move_legal(new_grid, piece, new_position, 6, 6) == false);
    REQUIRE(is_move_legal(new_grid, rotate_piece(piece, 1), new_position, 6, 6) == false);
    REQUIRE(is_move_legal(new_grid, rotate_piece(piece, 2), new_position, 6, 6) == false);
    REQUIRE(is_move_legal(new_grid, rotate_piece(piece, 3), new_position, 6, 6) == false);

    std::tie(success, new_grid, new_position) = place_piece_on_grid(new_grid, piece, new_position, rows, cols);
    REQUIRE(success == false);

    std::vector<std::array<int, 4>> pieces;
    pieces.push_back(piece);
    auto _result = get_valid_next_moves(new_grid, pieces, position, 6, 6);
    REQUIRE(_result.size() == 0);
}

TEST_CASE("Place piece on grid", "[factorial]") {
    int cols = 6; 
    int rows = 6; 
    std::array<int, 2> position = {0, 0};
    std::array<int, 4> piece = {0, 2, 3, 0}; 
    std::vector<int> grid = initialize_grid(rows, cols);

    std::vector<int> new_grid;
    std::array<int, 2> new_position;
    bool success;

    std::tie(success, new_grid, new_position) = place_piece_on_grid(grid, piece, position, rows, cols);
    REQUIRE(success == true);
    REQUIRE(new_grid[0 * cols * 4 + cols * 0 + 0] == 0);
    REQUIRE(new_grid[0 * cols * 4 + cols * 0 + 1] == 2);
    REQUIRE(new_grid[0 * cols * 4 + cols * 0 + 2] == 3);
    REQUIRE(new_grid[0 * cols * 4 + cols * 0 + 3] == 0);
    REQUIRE(grid[0 * cols * 4 + cols * 0 + 3] == EMPTY);
}

TEST_CASE("Get valid next moves", "[factorial]") {
    int cols = 6; 
    int rows = 6; 
    std::array<int, 2> position = {0, 0};
    std::array<int, 4> piece = {1, 2, 0, 0}; 
    std::vector<std::array<int, 4>> pieces;
    pieces.push_back(piece);
    std::vector<int> grid = initialize_grid(rows, cols);
    std::vector<std::pair<int, int>> result;

    result = get_valid_next_moves(grid, pieces, position, 6, 6);
    REQUIRE(result.size() == 1);
    REQUIRE(result[0].first == 0);
    REQUIRE(result[0].second == 1);
}

TEST_CASE("Find best position per hole tile", "[factorial]") {
    int cols = 4; 
    int rows = 4; 
    std::vector<Piece> pieces_b = initialize_pieces_backtracker(
            "test_fixtures/RandPuzzle4x4_01.txt");

    Piece **rotated_pieces = get_rotated_pieces_b(pieces_b);
    board  board;
    // 0/0   11/0    4/0    3/1
    // 8/3   12/2   15/2   10/1
    // 5/3   13/2   14/2    9/1
    // 2/3    7/2    6/2    1/2
    //
    //
    
    
    find_best_position_per_hole_tile (positions list_of_positions, board board, Piece ** rotated_pieces, int count_correct_edges);
    REQUIRE(pieces_b.size() == 16);
}
